### PR TITLE
Added base wrapper and wrapper like interface

### PR DESCRIPTION
**Pull request by nandi95**
linked #369

### CODE REVIEW

File: src/BaseWrapper.ts
Comment (nandi95) [inline src/BaseWrapper.ts:82]: This is an API change.

File: src/domWrapper.ts
Comment (nandi95) [inline src/domWrapper.ts:117]: This is an API change

**Comment by nandi95** [comment]
On the vueWrapper if it wasn't for the Omit 'exists' on the get method, then that could also be added to the WrapperLike interface.

**Comment by lmiller1990** [comment]
I think that Omit was intentional, maybe related: https://github.com/vuejs/vue-test-utils-next/issues/176

File: src/domWrapper.ts
Comment (lmiller1990) [reply src/domWrapper.ts:117]: This is a breaking change - you *must* await `setValue` now.

Although this change is probably good, I am not sure we want to introduce a breaking change. Let's see what others thinks.

Our docs cover when and why to use `await` extensively, this would require updating those a lot, too.

**Review by lmiller1990 — CHANGES_REQUESTED** [review]
I think the interface/base wrapper refactor seems fine, I am not sure on the breaking change.

Any chance we can remove the breaking change commit and discuss the elsewhere? Code improvements are great and always welcome, but any change to the API needs to be very carefully considered - we are in release candidate stage which generally means "no more breaking changes". 

We also have a pattern where we never wait for nextTick internally and leave that to the user (or another lib on top, like Testing Library), so this change would be in opposition to that.

File: src/BaseWrapper.ts
Comment (nandi95) [reply src/BaseWrapper.ts:82]: Actually this is fine, given this was async before and not accepting a callbeck to be passed to the next tick

File: src/domWrapper.ts
Comment (nandi95) [reply src/domWrapper.ts:117]: reverted

File: src/BaseWrapper.ts
Comment (nandi95) [reply src/BaseWrapper.ts:82]: It isn't actually. next tick returns `Promise<void>`

File: src/domWrapper.ts
Comment (afontcu) [reply src/domWrapper.ts:117]: 👍 Let's keep this reverted for now (so the PR does not introduce more changes than expected), but maybe making `setValue` async all the time is… a good thing? 🤔

File: src/interfaces/WrapperLike.ts
Comment (cexbrayat) [inline src/interfaces/WrapperLike.ts:11]: I think there should be the `get` method as well here

**Review by cexbrayat — COMMENTED** [review]
Looks good overall

Just 2 tiny feedback

**Comment by cexbrayat** [comment]
Oh and another note: we usually have commit message convention inspired by the Angular one:
- prefix by `feat`, `fix`, `refactor`, `chore`, `docs`
- use lowercase
- use imperative tense

These 2 commits could be just one with a message `refactor: introduce base wrapper`

File: src/interfaces/WrapperLike.ts
Comment (nandi95) [reply src/interfaces/WrapperLike.ts:11]: I'm not sure how to implement that given the Omit on the get see: https://github.com/vuejs/vue-test-utils-next/pull/372#issuecomment-776610237
I could have a second optional type argument but that felt messy.

File: src/errorWrapper.ts
Comment (nandi95) [reply src/errorWrapper.ts:1]: I'm nut sure why I did that.

File: src/interfaces/WrapperLike.ts
Comment (cexbrayat) [reply src/interfaces/WrapperLike.ts:11]: Hmm I think the Omit should be in both signatures, as `get` throws if the element is not there, so calling `exists` on it has no sense. If that's not the case that's a mistake? Or am I missing something @lmiller1990 ?

Comment (lmiller1990) [reply src/interfaces/WrapperLike.ts:11]: I think this is probably a mistake, yes

File: src/BaseWrapper.ts
Comment (lmiller1990) [inline src/BaseWrapper.ts:47]: nit: any reason this one is marked public? I think public is the default

**Comment by lmiller1990** [comment]
Thanks for reverting the API change, left one comment + @cexbrayat and we should be good... you are on fire 👍

**Review by lmiller1990 — APPROVED** [review]

**Comment by lmiller1990** [comment]
Ahhh one more tiny thing - all the files in this code-base are camel case - can we follow this convention?

**Review by lmiller1990 — APPROVED** [review]

**Comment by lmiller1990** [comment]
This is conflicted. I cannot merge it and I don't know why 🤔

**Pull request merged**
